File: main/scala/com/awesomesauce/minecraft/forge/openautomation/api/lasers/LaserPacket.scala
package com.awesomesauce.minecraft.forge.openautomation.api.lasers

import net.minecraft.world.World
import net.minecraftforge.common.util.ForgeDirection

trait LaserPacket {
  def arrive(world: World, x: Int, y: Int, z: Int, to: ForgeDirection): Boolean

  def split(amount: Int): Array[LaserPacket]

  def particleEffect: String = "reddust"
}

File: main/java/com/awesomesauce/minecraft/forge/openautomation/api/lasers/LaserPacket.java
package com.awesomesauce.minecraft.forge.openautomation.api.lasers;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;

public abstract class LaserPacket {
    NBTTagCompound compound = new NBTTagCompound();

    public abstract boolean arrive(World world, int x, int y, int z, ForgeDirection to);

    public abstract LaserPacket[] split(int amount);

    public String particleEffect() {
        return "reddust";
    }

    public NBTTagCompound getCompound() {
        return compound;
    }
}